### ISSUE TITLE
Deprecate internal transpilation layer

### ISSUE THREAD

**Issue by guybedford** [systemjs/systemjs#579]
Currently transpilation is entirely coupled to ES modules. This made sense previously, but is starting to reach its limits:
- What if we want to have scripts (not modules), which are written using new ES features, and we want to use the same transpiler?
- What happens when ES modules are available in browsers, and "format": "es" is the common scenario? Then we don't want transpilation at all.

The current plan here is:
- Ensure we have quality plugins for TypeScript, Traceur and Babel that all build and configure very easily supporting standard SystemJS and jspm workflows
- **Then and only then, entirely deprecate the internal transpilation layer of SystemJS (not the ES6 module loader polyfill though)**

This will be a very slow deprecation, focusing on advocating workflows not using the internal transpilation first.

**Comment by probins** [comment]
I'd agree that it makes sense to separate 'transpile' from 'ES6 module format', but ES6 support in browsers is progressing nicely at the moment - with the notable exception of import/export, which I don't see anybody working on much until the loader spec is finalised. So unless there's some action on that front soon, ISTM transpilation will increasingly only be needed for module syntax.

**Comment by guybedford** [comment]
It's also worth looking at having all transpilation run through plugins, and removing the transpilation entirely from System core. This would also entirely separate the module format from the transpiler.

This seems to be me the most appealing architecture, but my worry with this is that the ESnext transpilation is now coupled to its transpilation implementation - Babel / Traceur / TypeScript via:

```javascript
System.config({
  meta: {
    "app/*.js": { loader: 'babel', babelOptions: {...} }
  }
});
```

That is, you'll have `loader: 'babel'` all over the place to change instead of just a single `transpiler: 'babel'`. We could attempt to encourage a best-practice here in the getting started with something like:

```javascript
System.config({
  paths: {
    'esnext': 'path/to/babel.js'
  },
  meta: {
    '*': {
      babelOptions: {}
    }
    'app/*.js': {
      loader: 'esnext'
    }
  }
});
```

So that we maintain the same portability of the existing configuration. That may well make this approach tenable.

**Comment by guybedford** [comment]
I don't tend to think that the configuration portability should be an argument against good architecture here. Still tending towards separating all transpilation out into plugins.

**Comment by guybedford** [comment]
Updated the issue description with the current plan.

**Comment by guybedford** [comment]
It is still worth considering if the concept of some separate transpilation layer has merit.

Basically transpilation can be considered as two phases always. The first phase is bringing other languages to JavaScript. The second phase is compiling that JavaScript for the environment it is running in. JavaScript will always exist in its own future for some time, so separating these phases will allow other languages to transpile into the future of JS and not have to deal with that layer as well.

Potentially just through a transpiler option, this becomes a simple two-plugin chaining. Our `loader` plugin chains into our `transpiler` plugin. Both are written as plugins in the conventional sense of being loader hooks in userland code.

Something like:

```javascript
System.config({
  '*.js': {
    loader: 'hbs',
    transpiler: 'babel',
    babelOptions: { ... }
  }
});
```

Still on the fence somewhat but worth considering for these use cases.

**Comment by guybedford** [comment]
Such composition can still be enabled by making it very easy to use the Babel transpiler. For example:

```javascript
import {translate} from 'babel-plugin';
export function translate(load) {
  return translationFromOtherLang(load.source)
  .then(function(source) {
    return babelTranslate(source, load.metadata.babelOptions);
  });
}
```

The other composition use case is for bundling up transpilation options via plugins, this can be handled via an option factory like:

```javascript
import {optionFactory} from 'babel-plugin';
export default optionFactory({
  some: 'options'
});
```

Where the above options merge with any other `load.metadata.babelOptions` options.

These compositions work absolutely fine for all of these use cases. And also encourage convergence of transpilation on a single parse tree which is important for performance.

**Comment by mikz** [comment]
@guybedford would it still be possible to use transpiler + loader? For example babel + react-hot-api ?
I'm not sure if this is supported right now (#719). Would be the loader able to control whether it is executed first the transpiler or the loader?

**Comment by guybedford** [comment]
The loader would have to extend a base transpiler loader if it wants to enable transpilation.

**Comment by evanplaice** [comment]
What about letting users define a global implicit/explicit config

**Example of implicit:**

*config.js*
```
System.config({
  ...
  transpiler: "traceur",
  // transpileAll: true, defaults to true until standards are implemented
  traceurOptions: {
    "annotations": true,
  ...
  transpileExclude : [
    "semantic-ui",
    "jQuery",
    "redux"
  }
  ...
}
```
---------
**Example of explicit:**

*config.js*
```
System.config({
  ...
  transpiler: "traceur",
  transpileAll: false,
  traceurOptions: {
    "annotations": true,
  ...
  transpileInclude : [
    "angular2",
    "zone.js",
    "reflect-metadata"
  }
  ...
}
```
Whatever the solution, it shouldn't affect how modules are defined via `import`/`<script>`. Transpiler loading is an implementation detail that will likely change as the new standards are implemented.

-----
**Use case:**

A large media site spends months and x amount of dollars converting their 500 page site to use Angular2/React. Part of that process involved explicitly defining that modules use the transpiler in the source files.

Browsers finally catch up and ES6/ES7 is widely supported. They want to remove the transpiler requirement but making the change requires locating and editing all of the source files that depend on ES6/7 features.

-----
**Use Case 2:**

A dev team uses a package that just released a new major version that uses ES6/7 features (ie not yet supported by the browser).

Devs want to update their project to take full advantage of the latest release but enabling ES6/7 would require changing the source to enable transpilation (ex `import cool-lib { loader: "babel" }`). 

The site lacks 100% test coverage and one of the imports is missed breaking the production deployment.

------
Limiting configuration to the global `config.js` has two benefits for dev teams developing an existing codebase. First, It provides a 'safe' upgrade path to adopting libraries that use 'bleeding-edge' standards. Second, it provides a path of least resistance for teams that want to disable the transpiler piecemeal as standards are implemented in browsers.

This applies now to ES6/7 currently but it's important to consider how the future may look. Transpiling is not only used to provide polyfills for standards yet-to-be-implemented but also for languages that are a superset of Javascript. 

Standards are slow to be implemented and Javascript-like extension languages are becoming more common. Typescript and dart are good examples but asm.js is well on it's way to bridging C/C++ to the browsers.

-----
**Implementation:**

Checking for exceptions shouldn't too difficult to implement as dependencies are already mapped as a dictionary. Assuming O(1) searching for the dependency on import, searching for include/exclude should have the same characteristics.

Ideally, it would be beneficial to mark packages that require a transpiler in the registry and display a warning in the CLI on install. Except, that creates a maintenance issues with maintaining the registry and wouldn't apply to packages loaded from JSPM/Github.

----
**Aside:**
Personally, with no prior experience using a transpiler I was amazed at how easy it was able to use JSPM to start experimenting with Angular2 (ES7 not Typescript). Without it I probably would have worked through the pain of setting up grunt/gulp to automate the transpilation process but I'm sincerely glad I didn't have to.

JSPM not only benefits users by lowering the barrier. A growing user base greatly benefits the dev teams creating new Javascript language extensions.

**Comment by WorldMaker** [comment]
I was thinking about this the other night and about how essentially you want to represent a pipeline of operations/transformations. What if you want to convert things to an n-step process, beyond just two? On the one side, babel is trying to grow into being a configurable n-step processing pipeline and maybe the answer is that the transpiler parts of SystemJS just become the babel configuration parts? (Imagine a traceur plugin inside of babel, for instance. Prior to Babel 6 this was probably a good laugh, but again with the current direction babel is attempting to head it's seeming like more of a possibility...)

What I stumbled upon in thinking about some of these ideas was that this concept of n-step post-processing has a passing familiarity of sorts with the concept of Middleware in most server-side frameworks. I don't know how much of a useful metaphor that is (particularly because the Middleware concept itself can be fuzzy between different implementations in different frameworks), but maybe it does help to generalize the difference between a loader and a transpiler if you think of things as a stack of processing steps that is somewhat Middleware-esque.

**Comment by guybedford** [comment]
Deprecation warning added in https://github.com/systemjs/systemjs/commit/d3b615feb7a1285e6175e8085b0eccabc9693157, for removal in 0.17.

**Comment by sberney** [comment]
After this, if a plugin is enabled matching all "*.js", will the transpilation-needed recognition work pretty much as before?

And what about plugin chaining (just curious)?

**Comment by guybedford** [comment]
We will continue to support plugin -> ES module, and then a transpiler plugin for supporting ES modules. Chaining workflows are best done through a single plugin pipeline / extension process allowing a shared parse tree. Plugins aren't designed to chain individually.

**Comment by guybedford** [comment]
Implemented in [0.20.0-alpha.1](https://github.com/systemjs/systemjs/releases/tag/0.20.0-alpha.1).
